# Laboratorio de regresión - 3

## Significancia de factores

|                |   |
:----------------|---|
| **Nombre**     |Elisa Aguirre   |
| **Fecha**     2/3/26 |   |
| **Expediente**738894 |   |

Descarga el archivo de publicidad y carga los datos (Advertising.csv).

In [1]:
import pandas as pd 
import numpy as np
from sklearn.linear_model import LinearRegression
import matplotlib.pyplot as plt

In [4]:
df = pd.read_csv(r"C:\Users\elisa\Downloads\Advertising (1).csv")
df.head(10)

Unnamed: 0     TV  radio  newspaper  sales
0           1  230.1   37.8       69.2   22.1
1           2   44.5   39.3       45.1   10.4
2           3   17.2   45.9       69.3    9.3
3           4  151.5   41.3       58.5   18.5
4           5  180.8   10.8       58.4   12.9
5           6    8.7   48.9       75.0    7.2
6           7   57.5   32.8       23.5   11.8
7           8  120.2   19.6       11.6   13.2
8           9    8.6    2.1        1.0    4.8
9          10  199.8    2.6       21.2   10.6

**¿Hay alguna relación entre el presupuesto para publicidad y las ventas?**

Nuestra primera meta debe ser determinar si hay evidencia en los datos de que haya una asociación entre estas variables.

- ¿Por qué? ¿Qué resultaría si nos diéramos cuenta de la falta de relación entre el presupuesto de publicidad y las ventas?

Esto es importante porque, si existe una asociación, invertir en publicidad podría influir directamente en el nivel de ventas.
Si no existiera ninguna relación entre el presupuesto de publicidad y las ventas, entonces aumentar o disminuir la inversión publicitaria no tendría un efecto claro sobre los resultados. En ese caso, la publicidad no sería una herramienta útil para predecir o mejorar las ventas, y cualquier cambio observado podría ser por azar.

**¿Qué tan fuerte es esta relación?**
Asumiendo que existe esta relación, ¿nos sirve conocer el impacto que tiene invertir en publicidad en las ventas?

Asumiendo que sí existe una relación entre la publicidad y las ventas, es importante saber qué tan fuerte es dicha relación. Una relación fuerte indicaría que los cambios en el presupuesto de publicidad están asociados con cambios en las ventas.
Conocer esta relación es útil porque permite medir el impacto que tiene invertir en publicidad. De esta forma, las empresas pueden tomar decisiones más informadas sobre cuánto invertir y en qué medios, buscando maximizar las ventas y optimizar el uso de recursos.

**¿Cuáles medios están asociados con las ventas? ¿Qué tan grande es la asociación entre un medio específico y las ventas?**

Hay 3 medios distintos en los datos. ¿Sirve invertir en los 3? ¿Conviene más invertir sólo en uno?

**¿Qué tan seguros estamos de que podríamos predecir ventas futuras?**

**¿La relación es lineal?**

**¿Hay sinergia entre estos medios?**

Puede ser que gastar \\$50,000 en publicidad y otros \\$50,000 en radio es mejor opción que gastar \\$100,000 en televisión. A esto le llamamos efecto de interacción.

Podemos usar regresión lineal para responder todas estas preguntas.

Realiza una regresión lineal:

$$ \text{ventas} \approx \beta_0 + (\beta_1)(\text{TV})$$

In [6]:
X = df[["TV"]]
y = df["sales"]
modelo_tv = LinearRegression()
modelo_tv.fit(X, y)
beta_0 = modelo_tv.intercept_
beta_1 = modelo_tv.coef_[0]
print("Modelo de regresión lineal:")
print(f"ventas = {beta_0} + {beta_1}x")

Modelo de regresión lineal:
ventas = 7.032593549127693 + 0.047536640433019764x


### Verificando la precisión de nuestros coeficientes estimados

Recuerda que en el mundo real hay ruidos y errores de medición. Siempre se asume que la verdadera relación entre $X$ y $Y$ es $$Y = \beta_0 + \beta_1 X + \epsilon$$

Se asume que el término de error es independiente de $X$ (el error siempre es el mismo sin importar el valor de $X$). Este modelo describe a la *línea de regresión de la población*, que es la mejor aproximación de la verdadera relación entre $X$ y $Y$. Cuando usamos mínimos cuadrados encontramos la *línea de mínimos cuadrados*.

¿Cuál es la diferencia entre población y muestra?

La población se refiere al conjunto total de todos los elementos de interés que se desean estudiar, mientras que una muestra es solo una parte de esa población. La muestra se utiliza cuando no es posible o práctico analizar a toda la población.
En el contexto de la regresión, la población representa la relación verdadera entre las variables, mientras que la muestra es un subconjunto de datos que usamos para estimar dicha relación. Debido a esto, los resultados obtenidos a partir de una muestra pueden variar dependiendo de qué observaciones se incluyan.

¿Cuál crees que sea la diferencia entre hacer una regresión con todos los datos de la población y una muestra de ella?

Cuando se realiza una regresión utilizando todos los datos de la población, los coeficientes estimados describen exactamente la relación real entre las variables, sin error de muestreo. En este caso, no existe incertidumbre asociada a la estimación de los parámetros.
En cambio, al trabajar con una muestra, los coeficientes obtenidos son solo una aproximación de la relación verdadera. Estos estimadores pueden verse afectados por el tamaño de la muestra y por la variabilidad de los datos, lo que introduce incertidumbre y posibles errores en la estimación. Por esta razón, al usar muestras es importante evaluar la precisión de los coeficientes y reconocer que los resultados pueden cambiar si se toma una muestra diferente.


La línea de regresión de la población no se puede observar. El concepto de comparar estas líneas es una extensión natural del acercamiento estadístico estándar de usar información de una muestra para estimar características de una población grande.

Imagina que quieres encontrar la altura promedio de un mexicano $\mu$. Medir a todos y cada uno de los mexicanos en situaciones similares, con la misma regla, mismo operador, y otras incontables formas de minimizar la variación de la medida es una tarea imposible. Lo que podemos asumir es que $\hat{\mu} = \bar{y}$. La media poblacional y la media muestral son diferentes, pero la media muestral es usualmente un buen estimado.

De la misma manera, como no contamos con el 100% de la información para hacer una regresión, los coeficientes $\beta_0$ y $\beta_1$ son desconocidos. Podemos estimarlos usando mínimos cuadrados, encontrando $\hat{\beta_0}$ y $\hat{\beta_1}$. Puede que las muestras que tengamos en ese momento estén un poco por encima de la media, pero otras muestras en otro momento puede que estén debajo de la media. En general, esperamos que el promedio de las aproximaciones $\hat{\mu}$ aproxime a $\mu$.

Esto lleva a la pregunta: ¿qué tan cercanos son nuestros coeficientes estimados a los verdaderos coeficientes? Utilizamos el concepto de error estándar para evaluar esto.

$$ \text{Var}(\hat{\mu})=\text{SE}(\hat{\mu})^2 = \frac{\sigma^2}{n} $$

Donde $\sigma$ es la desviación estándar de cada una de las observaciones $y_i$ de $Y$. El error estándar nos dice la cantidad promedio que el estimado difiere del valor verdadero. Podemos ver en la fórmula que entre más observaciones tengamos el error se hace más pequeño. Las fórmulas para errores estándar de $\hat{\beta_0}$ y $\hat{\beta_1}$ son:

$$ \text{SE}(\hat{\beta_0})^2 = \sigma^2 [\frac{1}{n} + \frac{\bar{x}^2}{\sum_{i=1}^n (x_i - \bar{x})^2}]$$

$$ \text{SE}(\hat{\beta_1})^2 = \frac{\sigma^2}{\sum_{i=1}^n (x_i - \bar{x})^2}$$

$$ \sigma^2 = \text{Var}(\epsilon) = \text{RSE}^2 = \frac{\text{RSS}}{n-p}$$

Para que estas fórmulas sean validas asumimos que los errores $\epsilon_i$ tienen varianza común $\sigma^2$ y que no están correlacionados.

Calcula los errores estándar de los coeficientes

In [7]:
y_pred = modelo_tv.predict(X)
residuos = y - y_pred
n = len(y)
p = 2
RSS = np.sum(residuos**2)
sigma2 = RSS / (n - p)

In [9]:
promx = np.mean(x)
sumatoria = np.sum((x - promx)**2)
seb0 = np.sqrt(sigma2*((1/n)+(promx)**2/sumatoria))
seb1 = np.sqrt(sigma2/sumatoria)

In [11]:
b0 = modelo_tv.intercept_
b1 = modelo_tv.coef_

In [12]:
print("B0 =", b0)
print("B1 =", b1)
print("SE(B0) =", seb0)
print("SE(B1) =", seb1)

B0 = 7.032593549127693
B1 = [0.04753664]
SE(B0) = 0.4578429402734785
SE(B1) = 0.0026906071877968707


Estos errores se pueden usar para calcular intervalos de confianza. Un intervalo de confianza del $95\%$ se define como un rango de valores en el cuál se encuentra el desconocido valor verdadero con un $95\%$ de probabilidad.

Otra forma de verlo es que si tomamos muestras repetidas y construimos un intervalo de confianza para cada una, el $95\%$ de los intervalos creados van a contener el valor verdadero. Para la regresión el intervalo de confianza del $95\%$ toma la forma:

$$ \hat{\beta_j} \pm 2\text{SE}(\hat{\beta_j})$$

Calcula los intervalos de confianza para los coeficientes estimados:

In [13]:
IC_b0_inf = b0 - 2*seb0
IC_b0_sup = b0 + 2*seb0
print("IC 95% para B0:", IC_b0_inf, IC_b0_sup)

IC 95% para B0: 6.116907668580736 7.94827942967465


In [14]:
IC_b1_inf = b1 - 2*seb1
IC_b1_sup = b1 + 2*seb1
print("IC 95% para B1:", IC_b1_inf, IC_b1_sup)

IC 95% para B1: [0.04215543] [0.05291785]


Los errores estándar también se usan para realizar pruebas de hipótesis. La prueba de hipótesis más común es probar la hipótesis nula de:

$$ H_0: \text{No hay relación entre } X \text{ y } Y \ \ \ \ (\beta_1=0)$$

contra la hipótesis alternativa:
$$ H_0: \text{Hay alguna relación entre } X \text{ y } Y \ \ \ (\beta_1 \neq 0)$$

Explica con tus palabras el significado de la hipótesis nula y la hipótesis alternativa.

La hipótesis nula nos dice que no existe relación entre X y Y, es decir, que el coeficiente B1 es igual a cero. En este caso, cambiar el presupuesto en publicidad no tendría un efecto real sobre las ventas.
La hipótesis alternativa plantea que sí existe una relación entre X y Y, lo que significa que B1 es diferente de cero y que el presupuesto en publicidad sí tiene un impacto en las ventas.

Para probal la hipótesis nula debemos determinar si nuestro estimado $\hat{\beta_1}$ de $\beta_1$ está lo suficientemente alejado de cero para que podamos decir con confianza que este valor no es cero. 

¿Qué tan lejos? Depende de qué tanta confianza tengamos en el estimado encontrado. Si nuestro error estándar es pequeño y nuestro estimado está alejado de cero podríamos decir que hay muy poca probabilidad de que el valor verdadero sea 0. En cambio, si nuestro error estándar es grande y nuestro estimado está muy cerca de cero, entonces podrías ser que el valor verdadero sea cero y que no haya relación entre las variables.

Se calcula un *estadístico t* dado por
$$ t = \frac{\hat{\beta_j} - \mu}{\text{SE}(\hat{\beta_j})} $$

donde $\mu$ es el valor contra el que queremos probar.

Calcula el estadístico t para tus coeficientes estimados, usando como referencia la prueba de hipótesis.

In [15]:
t_b0 = (b0 - 0) / seb0
t_b1 = (b1 - 0) / seb1
print("t para B0:", t_b0)
print("t para B1:", t_b1)


t para B0: 15.360275174117543
t para B1: [17.6676256]


La distribución t tiene forma de campana y se parece bastante a la distribución normal cuando $n > 30$. Ya sólo es cuestión de calcular la probabilidad de observar cualquier número tal que su valor absoluto sea igual o mayor que el valor absoluto del estadístico t calculado. En otras palabras:
$$ P(|x| \geq |t|) $$

A esta probabilidad la llamamos *p-value*. Un *p-value* pequeño indica que es poco probable que exista por puro azar una relación significativa entre predictor y respuesta, en caso de que no haya una asociación real entre predictor y respuesta. En otras palabras, el *p-value* te dice la probabilidad de que parezca que hay relación cuando no la hay.

Si el *p-value* es pequeño, inferimos que sí hay una asociación entre el predictor y la respuesta, y **rechazamos la hipótesis nula**.
  

¿Qué tan pequeño? Depende de la aplicación. Un valor muy común es del $5\%$.

Utiliza el siguiente código para calcular el *p-value* para tus coeficientes

`from scipy import stats`

`p_bj = 2*(1 - stats.t.cdf(np.abs(t_bj), n-p))`

In [16]:
from scipy import stats
p_b0 = 2 * (1 - stats.t.cdf(np.abs(t_b0), n - p))
p_b1 = 2 * (1 - stats.t.cdf(np.abs(t_b1), n - p))
print("p-value para B0:", p_b0)
print("p-value para B1:", p_b1)

p-value para B0: 0.0
p-value para B1: [0.]


¿Se rechaza la hipótesis nula? ¿Qué significa?

Sí, se rechaza la hipótesis nula. El p-value obtenido para el coeficiente B1 es prácticamente cero y menor al nivel de significancia del 5%. Esto indica que es muy poco probable observar un resultado como el obtenido si en realidad no existiera una relación entre el presupuesto en publicidad y las ventas.
Por lo tanto, se concluye que existe una relación estadísticamente significativa entre el gasto en publicidad (TV) y las ventas.

Realiza otras dos regresiones. Ya tienes hecha la regresión de ventas dado el gasto en publicidad de TV. Realiza la regresión para gastos en radio y gastos en periódico. Organiza las respuestas para que debajo de esta celda se tenga:
- Título de regresión
- Coeficientes estimados
- Errores estándar de los coeficientes
- Intervalos de confianza
- Estadísticos t
- p-values
- Observaciones

### Regresión lineal: Ventas vs Radio


In [17]:
# Regresión Ventas vs Radio
X = df[["radio"]]
y = df["sales"]
modelo_radio = LinearRegression()
modelo_radio.fit(X, y)
# Coeficientes
b0 = modelo_radio.intercept_
b1 = modelo_radio.coef_
# Predicciones y residuos
y_pred = modelo_radio.predict(X)
residuos = y - y_pred
# Parámetros
n = len(y)
p = 2
RSS = np.sum(residuos**2)
sigma2 = RSS / (n - p)
# Cálculos auxiliares
x = X["radio"].values
promx = np.mean(x)
sumatoria = np.sum((x - promx)**2)
# Errores estándar
se_b0 = np.sqrt(sigma2 * ((1/n) + (promx**2)/sumatoria))
se_b1 = np.sqrt(sigma2 / sumatoria)
# Intervalos de confianza
IC_b0 = (b0 - 2*se_b0, b0 + 2*se_b0)
IC_b1 = (b1 - 2*se_b1, b1 + 2*se_b1)
# Estadísticos t
t_b0 = b0 / se_b0
t_b1 = b1 / se_b1
# p-values
p_b0 = 2 * (1 - stats.t.cdf(np.abs(t_b0), n - p))
p_b1 = 2 * (1 - stats.t.cdf(np.abs(t_b1), n - p))
# Resultados
print("Regresión Ventas vs Radio")
print("Coeficientes estimados:")
print("B0 =", b0)
print("B1 =", b1)
print("Errores estándar:")
print("SE(B0) =", se_b0)
print("SE(B1) =", se_b1)
print("Intervalos de confianza 95%:")
print("B0:", IC_b0)
print("B1:", IC_b1)
print("Estadísticos t:")
print("t(B0) =", t_b0)
print("t(B1) =", t_b1)
print("p-values:")
print("p(B0) =", p_b0)
print("p(B1) =", p_b1)

Regresión Ventas vs Radio
Coeficientes estimados:
B0 = 9.311638095158283
B1 = [0.20249578]
Errores estándar:
SE(B0) = 0.5629004962718711
SE(B1) = 0.020411306360090635
Intervalos de confianza 95%:
B0: (np.float64(8.185837102614542), np.float64(10.437439087702025))
B1: (array([0.16167317]), array([0.2433184]))
Estadísticos t:
t(B0) = 16.542245311258217
t(B1) = [9.92076547]
p-values:
p(B0) = 0.0
p(B1) = [0.]


**Observaciones:**  
La regresión de ventas respecto al gasto en radio muestra una relación positiva y estadísticamente significativa. El coeficiente B1 es positivo (≈ 0.202) y su p-value es prácticamente cero, por lo que se rechaza la hipótesis nula de no relación. Además, el intervalo de confianza al 95% para B1 no contiene al cero, lo que refuerza la evidencia de que el gasto en radio tiene un impacto significativo en las ventas.

### Regresión lineal: Ventas vs Periódico


In [18]:
# Regresión Ventas vs Periódico
X = df[["newspaper"]]
y = df["sales"]
modelo_news = LinearRegression()
modelo_news.fit(X, y)
# Coeficientes
b0 = modelo_news.intercept_
b1 = modelo_news.coef_
# Predicciones y residuos
y_pred = modelo_news.predict(X)
residuos = y - y_pred
# Parámetros
n = len(y)
p = 2
RSS = np.sum(residuos**2)
sigma2 = RSS / (n - p)
# Cálculos auxiliares
x = X["newspaper"].values
promx = np.mean(x)
sumatoria = np.sum((x - promx)**2)
# Errores estándar
se_b0 = np.sqrt(sigma2 * ((1/n) + (promx**2)/sumatoria))
se_b1 = np.sqrt(sigma2 / sumatoria)
# Intervalos de confianza
IC_b0 = (b0 - 2*se_b0, b0 + 2*se_b0)
IC_b1 = (b1 - 2*se_b1, b1 + 2*se_b1)
# Estadísticos t
t_b0 = b0 / se_b0
t_b1 = b1 / se_b1
# p-values
p_b0 = 2 * (1 - stats.t.cdf(np.abs(t_b0), n - p))
p_b1 = 2 * (1 - stats.t.cdf(np.abs(t_b1), n - p))
# Resultados
print("Regresión Ventas vs Periódico")
print("Coeficientes estimados:")
print("B0 =", b0)
print("B1 =", b1)
print("Errores estándar:")
print("SE(B0) =", se_b0)
print("SE(B1) =", se_b1)
print("Intervalos de confianza 95%:")
print("B0:", IC_b0)
print("B1:", IC_b1)
print("Estadísticos t:")
print("t(B0) =", t_b0)
print("t(B1) =", t_b1)
print("p-values:")
print("p(B0) =", p_b0)
print("p(B1) =", p_b1)

Regresión Ventas vs Periódico
Coeficientes estimados:
B0 = 12.351407069278162
B1 = [0.0546931]
Errores estándar:
SE(B0) = 0.6214201876094345
SE(B1) = 0.016575721876358165
Intervalos de confianza 95%:
B0: (np.float64(11.108566694059293), np.float64(13.59424744449703))
B1: (array([0.02154165]), array([0.08784454]))
Estadísticos t:
t(B0) = 19.87609562024895
t(B1) = [3.29959074]
p-values:
p(B0) = 0.0
p(B1) = [0.0011482]


**Observaciones:**  
En la regresión de ventas respecto al gasto en periódico también se observa una relación positiva, aunque menor que en el caso del radio y la televisión. El coeficiente B1 es positivo (≈ 0.055) y su p-value es menor al 5%, por lo que se rechaza la hipótesis nula. Sin embargo, el efecto estimado es más pequeño y el intervalo de confianza es menor, lo que sugiere que el impacto del gasto en periódico sobre las ventas es más limitado.

## Regresión lineal múltiple

En lugar de hacer una regresión para cada factor independiente, quizás se puede extender el modelo para que tenga varios factores dentro:

$$ Y = \beta_0 + \beta_1 X_1 + \beta_2 X_2 + ... + \beta_p X_p + \epsilon $$

Para nuestro ejemplo de publicidad:

$$ \text{sales} = \beta_0 + \beta_1 (\text{TV}) + \beta_2 (\text{radio}) + \beta_3 (\text{newspaper}) + \epsilon $$

Utiliza la librería `statsmodels` para realizar la regresión. Por defecto la librería statsmodels no toma en cuenta el intercepto ($\beta_0$), por lo que se tendrá que agregar una columna de unos de tamaño *n* a la matriz X.

`import statsmodels.api as sm`

`ols = sm.OLS(Y, X)`

`results = ols.fit()`

`results.summary()`

In [20]:
import statsmodels.api as sm
Y = df["sales"]
X = df[["TV", "radio", "newspaper"]]
X = sm.add_constant(X)
ols = sm.OLS(Y, X)
results = ols.fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  sales   R-squared:                       0.897
Model:                            OLS   Adj. R-squared:                  0.896
Method:                 Least Squares   F-statistic:                     570.3
Date:                Tue, 03 Feb 2026   Prob (F-statistic):           1.58e-96
Time:                        16:47:35   Log-Likelihood:                -386.18
No. Observations:                 200   AIC:                             780.4
Df Residuals:                     196   BIC:                             793.6
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          2.9389      0.312      9.422      0.000       2.324       3.554
TV             0.0458      0.001     32.809      0.000       0.043       0.049
radio          0.1885      0.009     21.893      0.000       0.172       0.206
newspaper     -0.0010      0.006     -0.177      0.860      -0.013       0.011
==============================================================================
Omnibus:                       60.414   Durbin-Watson:                   2.084
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              151.241
Skew:                          -1.327   Prob(JB):                     1.44e-33
Kurtosis:                       6.332   Cond. No.                         454.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

### Regresión lineal múltiple
y = 2.9389 + 0.0458TV + 0.1885radio - 0.0010newspaper

¿Qué diferencias puedes observar entre los *p-values* de una regresión múltiple y los encontrados en las regresiones simples? ¿Por qué crees que existen estas diferencias?

Al comparar los p-values de las regresiones simples con los de la regresión múltiple, se observa que en las regresiones simples cada variable se analiza de manera individual, por lo que su p-value refleja la relación con las ventas sin considerar la influencia de los otros medios.
En la regresión lineal múltiple, los p-values miden el efecto de cada variable manteniendo constantes las demás. Esto explica por qué variables como *newspaper* pueden resultar significativas en una regresión simple, pero dejar de serlo en la múltiple, ya que su efecto se ve absorbido por la relación que tiene con otros medios como televisión y radio.
Estas diferencias se deben a que la regresión múltiple permite controlar la redundancia y la interacción entre variables explicativas, identificando cuáles tienen un impacto real e independiente sobre las ventas.


## Referencia

James, G., Witten, D., Hastie, T., Tibshirani, R.,, Taylor, J. (2023). An Introduction to Statistical Learning with Applications in Python. Cham: Springer. ISBN: 978-3-031-38746-3

## Addendum

Para calcular los *p-values* de los parámetros sin `statsmodels`:
1. Calcular RSS
2. Calcular RSE
3. `var_beta = np.linalg.inv(X.T @ X) * rse**2` (X con columna de intercepto)
4. `std_beta = np.sqrt(var_beta.diagonal())` El orden de los valores corresponde al orden de los factores en las columnas de la matriz $X$.
5. Calcular *estadístico t*
6. Calcular *p-value*